exchange/exchange_test.go

```diff
@@ -1196,6 +1196,238 @@ func TestCategoryDedupe(t *testing.T) {
+	}
+	targ4 := map[string]string{
+		"hb_pb":         "12.00",
+		"hb_pb_cat_dur": "12.00_30s",
+	}
+	targ5 := map[string]string{
+		"hb_pb":         "12.00",
+		"hb_pb_cat_dur": "12.00_auto_30s",
+	}
+
+	bid1_1 := pbsOrtbBid{&openrtb.Bid{
+		Ext: json.RawMessage(`{"appnexus":{"deal_priority":5}}`),
+	}, "video", targ1, &openrtb_ext.ExtBidPrebidVideo{}}
+	bid1_2 := pbsOrtbBid{&openrtb.Bid{
+		Ext: json.RawMessage(`{"appnexus":{"deal_priority":6}}`),
+	}, "video", targ2, &openrtb_ext.ExtBidPrebidVideo{}}
+	bid1_3 := pbsOrtbBid{&openrtb.Bid{
+		Ext: json.RawMessage(`{"appnexus":{"deal_priority":7}}`),
+	}, "video", targ3, &openrtb_ext.ExtBidPrebidVideo{}}
+	bid1_4 := pbsOrtbBid{&openrtb.Bid{
+		Ext: json.RawMessage(`{"appnexus":{"deal_priority":8}}`),
+	}, "video", targ4, &openrtb_ext.ExtBidPrebidVideo{}}
+	bid1_5 := pbsOrtbBid{&openrtb.Bid{}, "video", targ5, &openrtb_ext.ExtBidPrebidVideo{}}
+
+	auc := &auction{
+		winningBidsByBidder: map[string]map[openrtb_ext.BidderName]*pbsOrtbBid{
+			"imp_id1": {
+				bidderName1: &bid1_1,
+			},
+			"imp_id2": {
+				bidderName2: &bid1_2,
+			},
+			"imp_id3": {
+				bidderName3: &bid1_3,
+			},
+			"imp_id4": {
+				bidderName4: &bid1_4,
+			},
+			"imp_id5": {
+				bidderName5: &bid1_5,
+			},
+		},
+	}
+
+	applyDealSupport(bidRequest, auc)
+
+	assert.Equal(t, "tier5_movies_30s", auc.winningBidsByBidder["imp_id1"][bidderName1].bidTargets["hb_pb_cat_dur"], "hb_pb_cat_dur was not updated correctly")
+	assert.Equal(t, "second6_15s", auc.winningBidsByBidder["imp_id2"][bidderName2].bidTargets["hb_pb_cat_dur"], "hb_pb_cat_dur was not updated correctly")
+	assert.Equal(t, "12.00_medicine_30s", auc.winningBidsByBidder["imp_id3"][bidderName3].bidTargets["hb_pb_cat_dur"], "hb_pb_cat_dur should not have been changed")
+	assert.Equal(t, "12.00_30s", auc.winningBidsByBidder["imp_id4"][bidderName4].bidTargets["hb_pb_cat_dur"], "hb_pb_cat_dur should not have been changed")
+	assert.Equal(t, "12.00_auto_30s", auc.winningBidsByBidder["imp_id5"][bidderName5].bidTargets["hb_pb_cat_dur"], "hb_pb_cat_dur should not have been changed")
+}
+
+func TestGetDealTiers(t *testing.T) {
+	bidRequest := &openrtb.BidRequest{
+		ID: "some-request-id",
+		Imp: []openrtb.Imp{
+			{
+				ID:  "imp_id1",
+				Ext: json.RawMessage(`{"validbase": {"dealTier": {"minDealTier": 5, "prefix": "tier"}, "placementId": 10433394}}`),
+			},
+			{
+				ID:  "imp_id2",
+				Ext: json.RawMessage(`{"validmultiple1": {"dealTier": {"minDealTier": 5, "prefix": "tier"}, "placementId": 10433394}, "validmultiple2": {"dealTier": {"minDealTier": 5, "prefix": "tier"}, "placementId": 10433394}}`),
+			},
+			{
+				ID:  "imp_id3",
+				Ext: json.RawMessage(`{"nodealtier": {"placementId": 10433394}}`),
+			},
+			{
+				ID:  "imp_id4",
+				Ext: json.RawMessage(`{"onedealTier1": {"placementId": 10433394}, "onedealTier2": {"dealTier": {"minDealTier": 5, "prefix": "tier"}, "placementId": 10433394}}`),
+			},
+		},
+	}
+
+	filledDealTier := DealTier{
```

Comment:
I notice getDealTiers() is saving the address of a variable defined in its for loop, which might work ok in Go, but I think these tests should make sure a different map is used for each imp instead of all imps pointing to the same map. For example if "imp_id4" used "validmultiple1" as the first bidder name then that would be enough ensure it's not overwriting the same map entry that "imp_id2" is using (since imp2 should be filled and imp4 should be empty).

Reply:
Per our discussion, updated the unit test here to reuse a bidder ("validbase" in this case) to verify that the map is pointing to different deal tier objects.